stablehlo/conversions/tosa/transforms/StablehloQuantLegalizeToTosaRescale.cpp

```diff
@@ -0,0 +1,257 @@
+/* Copyright 2024 OpenXLA Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <memory>
+#include <utility>
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Quant/QuantOps.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Dialect/Tosa/Utils/ConversionUtils.h"
+#include "mlir/Dialect/Tosa/Utils/QuantUtils.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "stablehlo/conversions/tosa/transforms/Passes.h"
+#include "stablehlo/dialect/StablehloOps.h"
+
+#define PASS_NAME "stablehlo-quant-legalize-to-tosa-rescale"
+#define DEBUG_TYPE PASS_NAME
+
+using namespace mlir;
+using namespace mlir::tosa;
+
+namespace mlir {
+namespace tosa {
+
+#define GEN_PASS_DEF_STABLEHLOQUANTLEGALIZETOTOSARESCALEPASS
+#include "stablehlo/conversions/tosa/transforms/Passes.h.inc"
+
+namespace {
+
+struct StablehloQuantLegalizeToTosaRescalePass
+    : impl::StablehloQuantLegalizeToTosaRescalePassBase<
+          StablehloQuantLegalizeToTosaRescalePass> {
+  void runOnOperation() final;
+
+  LogicalResult initialize(MLIRContext* ctx) override;
+
+ private:
+  FrozenRewritePatternSet patterns;
+};
+
+#define DECL_CONVERT_OP(stablehlo_op)                                        \
```

Comment:
I could go either way on macros like this, but tend to prefer avoiding them -- IMO they inhibit readability/searchability at the cost of ~2 LoC difference per op (compating op decl+impl to just inline struct impl).

We tend to not need to declare patterns, can just implement the struct inline in an anonymous namespace and add to the pattern set.

```cpp
DECL_CONVERT_OP(Add);

LogicalResult ConvertStablehloQuantAddOp::matchAndRewrite(
    stablehlo::AddOp op, PatternRewriter& rewriter) const {
  return matchAndRewriteAddSub<stablehlo::AddOp>(op, rewriter);
}

vs

struct ConvertStablehloQuantAddOp : public OpRewritePattern<stablehlo::AddOp> {
  using OpRewritePattern::OpRewritePattern;
  LogicalResult matchAndRewrite(stablehlo::AddOp op,
                                PatternRewriter& rewriter) const override {
    return matchAndRewriteAddSub<stablehlo::AddOp>(op, rewriter);
  }
}
```

Reply:
fixed